File: Solution/剑指Offer06.从尾到头打印链表/剑指Offer06.从尾到头打印链表.py
'''
Descroption: 剑指 Offer 06. 从尾到头打印链表
Author: EmoryHuang
Date: 2021-06-28 17:03:00
解题思路:
简单题，reverse, 递归, 栈
'''


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # reverse
    def reversePrint(self, head: ListNode) -> List[int]:
        ans = []
        while head != None:
            ans.append(head.val)
            head = head.next
        return ans[::-1]

    # 递归
    # def reversePrint(self, head: ListNode) -> List[int]:
    #     if head is None:
    #         return []
    #     return self.reversePrint(head.next) + [head.val]

File: Solution/剑指Offer06.从尾到头打印链表/剑指Offer06.从尾到头打印链表.java
/*
 * @Descroption: 剑指 Offer 06. 从尾到头打印链表
 * @Author: EmoryHuang
 * @Date: 2023-07-24 09:43:30
 * @Method:
 * 递归
 */

import java.util.*;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public int[] reversePrint(ListNode head) {
        return dfs(head, 0);
    }

    int[] dfs(ListNode node, int size) {
        if (node == null) {
            return new int[size];
        }
        int[] nums = dfs(node.next, size + 1);
        nums[nums.length - size - 1] = node.val;
        return nums;
    }
}